### PR TITLE
patch for CM 0.10.763 + minor improvements

### PR DESCRIPTION
**Pull request by MrKingfisher**
was never a fan of having js mixed with dll plugins so i created some logic to create a folder and added ofc some checking for if  folder exist blah blah, and also let it move all current scripts inside of plugins folder to be moved into Plugins\CM-JsScripts,
and a 1 time created js script when folder is created to greet and inform when they click on it.

explorer_9xt82RfeKe

### CODE REVIEW

**Comment by MrKingfisher** [comment]
tho i am only novice at c# so please do review it and so on

**Comment by MrKingfisher** [comment]
also i think Unity.TextMeshPro.Editor is gone from reference Assembly,
didn't include the updated csproj as it bricked on me

File: Checks/ExternalJS.cs [hunk 1]
```diff
@@ -159,6 +161,8 @@ private void LoadJS()
             Debug.LogWarning($"Error loading {fileName}\n{jse.Message}");
         }
     }
+    private void startEngine() { 
+    }
```
Comment (XAce1337manX) [inline Checks/ExternalJS.cs]: Does this do anything?

File: .gitignore [hunk 2]
```diff
@@ -2,3 +2,5 @@ obj/
 bin/
 .vs/
 packages/
+/ErrorChecker.csproj
+/ErrorChecker.csproj
```
Comment (XAce1337manX) [inline .gitignore]: Probably shouldn't add these to the gitignore.

File: CMJS.cs [hunk 3]
```diff
@@ -28,25 +28,45 @@ public class CMJS
...
-        new EmptyCheck()
+        // Handle yourself...
     };
     private CheckResult errors;
     private UI ui;
     private AudioTimeSyncController atsc;
     private int index = 0;
     private bool movedAfterRun = false;
 
+    private const string JsScriptDirectory = "CM-JsScripts";
```
Comment (XAce1337manX) [inline CMJS.cs]:
```suggestion
    private const string JsScriptDirectory = "CM-JS-Scripts";
```
The plugin is `CM-JS` so lowercase feels off imo.

Also you probably want to put these constants somewhere or expose it so `ExternalJS.cs` can reference the same constant.

File: CMJS.cs [hunk 4]
```diff
@@ -28,25 +28,45 @@ public class CMJS
...
+    private const string DefaultJsScript = "cmjsupdate.js";
+
     [Init]
     private void Init()
     {
+        // Standardsize directory creation for js scripts.
```
Comment (XAce1337manX) [inline CMJS.cs]: Comment typo

File: CMJS.cs [hunk 5]
```diff
@@ -28,25 +28,45 @@ public class CMJS
...
+        EnsureJsScriptDirectoryExists();
         SceneManager.sceneLoaded += SceneLoaded;
 
         string assemblyFolder = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+        string jsPluginsFolder = Path.Combine(assemblyFolder, JsScriptDirectory);
+        // depreciate load of JS scripts directly from plugins folder directly,
```
Comment (XAce1337manX) [inline CMJS.cs]: Probably meant `deprecate`

File: CMJS.cs [hunk 6]
```diff
@@ -28,25 +28,45 @@ public class CMJS
...
-            checks.Add(new ExternalJS(Path.GetFileName(file)));
+        {   // Yes we relocate them to new folder right before loading them
+            string fileName = Path.GetFileName(file);
+            // Construct the destination file path
+            string destFilePath = Path.Combine(jsPluginsFolder, fileName);
+            try
+            {
+                File.Move(file, destFilePath);
+            }
+            catch (IOException IOError) { Debug.LogError("IOException: thrown while trying to relocate script from: " + file + " To " + jsPluginsFolder + " <- this foler");
```
Comment (XAce1337manX) [inline CMJS.cs]:
```suggestion
            catch (IOException IOError) { Debug.LogError("IOException: thrown while trying to relocate script from: " + file + " to " + jsPluginsFolder + " <- this folder");
```

**Review by XAce1337manX — CHANGES_REQUESTED** [review]
Looks ok. Needs some cleanup.

File: Checks/ExternalJS.cs (hunk 1 above)
Comment (MrKingfisher) [reply Checks/ExternalJS.cs]: nope. was gonna add some then i was like fuck it and i forgot to remove it, TODO today

File: CMJS.cs (hunk 5 above)
Comment (MrKingfisher) [reply CMJS.cs]: yep

**Comment by MrKingfisher** [comment]
will be done in less than 6 hrs

**Comment by MrKingfisher** [comment]
found a minor issue or perhaps its more of a QoL feature.
if you load in scripts while scromapper is active it won't load them before a restart,
was working on a button to reload javascripts, it did work but not properly so it was discarded

**Review by XAce1337manX — APPROVED** [review]

**Pull request merged**